server/api.js

```diff
@@ -19,74 +16,129 @@ const router = new Router();
-		} else {
-			client.close();
-			return res.send(400);
+		// check if the id is valid if not -> 404
+		if (!mongodb.ObjectID.isValid(req.params.id)) {
+			return res.send(404);
 		}
-		const searchObject = { _id: id };
-
-		collection.findOne(searchObject, function (
-			error,
-			student
-		) {
-			res.send(error || student);
-			client.close();
+
+		// define id (mongodb.ObjectID)
+		const id = new mongodb.ObjectID(req.params.id);
+
+		const queryObject = { _id: id };
+
+		collection.findOne(queryObject, (error, result) => {
+			if (error) {
+				return res.status(500).send(error);
+			}
+
+			// if no record ->
+			if (!result) {
+				return res.sendStatus(404);
+			}
+
+			// if record -> send the data (200 default)
+			return res.send(result);
 		});
+
 	});
-});
 
-router.post("/", (req, res, next) => {
-	const client = getClient();
-	client.connect((err) => {
-		if (err) {
-			return next(err);
+	// To post new feedback or area of focus
+	router.put("/students/:id", (req, res) => {
+		const collection = db.collection("StudentDenormalizedData");
+
+		const data =  req.body ;
+		// validation should happen here
+
+		// check if the id is valid if not -> 404
+		if (!mongodb.ObjectID.isValid(req.params.id)) {
+			return res.send(404);
 		}
-		const db = client.db("feedback-tracker");
-		const collection = db.collection("feedback");
-		const data = req.body;
 
-		collection
-			.insertOne(data, (error, result) => {
-			// if everything is not ok -> send error response (500)
-				if (error) {
-					console.log(error);
-					return res.sendStatus(500);
-				}
-				// if everything is ok -> send returned record (a bit tricky to find it...)
-				return res.status(201).send(result.ops[0]);
-			});
+		// define id (mongodb.ObjectID)
+		const id = new mongodb.ObjectID(req.params.id);
+		const queryObject = { _id: id };
+		// DON'T send back the original record, send back the UPDATED record
+		const options = { returnOriginal: false };
+		const sendErrorOrResult = (error, result) => {
+			if (error) {
+				return res.status(500).send(error);
+			}
+			return res.send(result.value); // result.value === result.ops[0]
+		};
+
+		if (data.hasOwnProperty("level")) {
+			const level = data.level;
+			const message = data.message;
+
+			level === "toWorkOn"
+				? collection.update(
+					queryObject,
+					{ $push:
+					{ "areasOfFocus.toWorkOn": message } },
+					options,
+					sendErrorOrResult
+				) : level === "okayAt"
+					? collection.update(
+						queryObject,
+						{ $push:
+					{ "areasOfFocus.okayAt": message } },
+						options,
+						sendErrorOrResult
+					) : level === "goodAt"
+						? collection.update(
+							queryObject,
+							{ $push:
+					{ "areasOfFocus.goodAt": message } },
+							options,
+							sendErrorOrResult
+						)
+						: null;
+			return 1;
+		}
+
+		// if there are missing mandatory properties then return an error message to the API user and stop
+		function findMissingFields(data, mandatoryFields) {
```

Comment:
Is mandatoryFields a keyword or I could not find it declared above?

Reply:
No it is not a keyword, just a parameter.